colossalai/gemini/tensorful_state.py

```diff
@@ -0,0 +1,172 @@
+        if self._state == TensorState.FREE:
+            # when the state is free, payload should be None
+            assert self._payload is None, f"payload has to None if state is {self._state}"
+        else:
+            # otherwise, payload should not be None
+            assert self._payload is not None, f"payload can't be None if state is {self._state}"
+            StatefulTensor.trans_state_update(tensor, TensorState.FREE, state)
+
+    def data_ptr(self):
+        if self._payload is None:
+            return 0  # if a tensor has no storage, 0 should be returned
+        return self._payload.data_ptr()
+
+    def set_null(self) -> None:
+        # notice that free stateful tensor do not need to become null again
+        if self.state != TensorState.FREE:
+            StatefulTensor.trans_state_update(self.payload, self.state, TensorState.FREE)
+            self._state = TensorState.FREE
+            self._payload = None
+
+    def is_null(self) -> bool:
+        if self.state == TensorState.FREE:
+            # check sanity here
+            assert self.payload is None
+            return True
+        return False
+
+    def trans_state(self, state: TensorState) -> None:
+        if self.state == TensorState.FREE:
+            # free stateful tensor can't change state
+            assert state == TensorState.FREE, "Free stateful tensor can't change to other states"
+            return
+
+        StatefulTensor.trans_state_update(self.payload, self.state, state)
+        self._state = state
+        if state == TensorState.FREE:
+            self._payload = None
+
+    def move_to(self, device: Union[torch.device, int]):
+        assert self.state is not TensorState.FREE, "Can't move free stateful tensor"
+
+        if not isinstance(device, torch.device):
+            to_device = torch.device('cuda', device)
+        else:
+            to_device = device
+
+        from_device_type = self.payload.device.type
+        if from_device_type == to_device.type:
+            # from device == to device
+            return
+
+        # update manager's information
+        StatefulTensor.trans_device_update(self.payload, from_device_type, to_device.type, self.state)
+        self.payload.data = self.payload.data.to(to_device)
+
+    def copy_payload(self, tensor) -> None:
+        self._payload.view(-1).copy_(tensor.view(-1))
+
+    def reset_payload(self, tensor) -> None:
+
+        assert tensor is not None, "Can't reset None as the payload of stateful tensor, please use set_null() instead"
+
+        if self.payload is not None:
+            # if current state isn't FREE, release old payload
+            StatefulTensor.trans_state_update(self.payload, self.state, TensorState.FREE)
+        else:
+            # otherwise, set the state to HOLD for new payload
+            self._state = TensorState.HOLD
+        del self._payload
+
+        # record new payload
+        StatefulTensor.trans_state_update(tensor, TensorState.FREE, self.state)
```

Comment:
We must call this function when the payload is newed or deleted.
Therefore trans_state_update is coupled with reset_payload and move_to.
Here are some negative effects.
First, it is not applicable to the other place. Because you have to keep in mind that the tensor is exactly allocated or deleted.
Or some others should not use the function. It is dangerous and will mess up your plan.

Reply:
I'm gonna change them to private functions.